implement dynamic bucket order in DDP

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35137 implement dynamic bucket order in DDP**

bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

zhaojuanmao committed:
implement dynamic bucket order in DDP

bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -795,16 +872,21 @@ std::vector<std::vector<size_t>> compute_bucket_assignment_by_size(
 
+    auto param_indice = i;
+    if (!tensor_indices.empty()) {
+      param_indice = tensor_indices[i];
```
indice -> index

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -186,11 +189,21 @@ class Reducer {
 
+// tensor_indices stores the parameter tensor indices of "tensors" in
+// orginal model.parameters() list, if it is empty, then the parameter tensor
+// indices of "tensors" are the indices in "tensor" list.
```
when can these two (`tensors` and `model.parameters()`) be different?

mrshenli replied:
Oh I see, it is also called in `rebuildBuckets `

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +753,59 @@ void Reducer::finalize_backward() {
+
+  // Rebuilt bucket if this is the first time to rebuilt
+  if (!rebuilt_tensors_.empty()) {
+    rebuildBuckets();
```
Since we don't need this until the next backward, does it worth it to make the rebuild process async?

mrshenli replied:
Since this will launch communication (broadcast), we'd better do it async.

zhaojuanmao replied:
Good idea! I think we can make it async for find_unused_parameter = true in the future, as it could rebuild buckets much more times.  For  find_unused_parameter = false, since we only rebuild once, and only broadcast indices not tensors, it should be fast enough.  I will add a TODO comment here. 

How do you think?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -282,6 +290,21 @@ void Reducer::autograd_hook(VariableIndex index) {
+      index.replica_index == 0) {
+    rebuilt_tensors_.push_back(
+        replicas_[index.replica_index][index.variable_index]);
+    rebuilt_param_indices_.push_back(index.variable_index);
```
How do we guarantee this order is consistent across all DDP processes?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -282,6 +290,21 @@ void Reducer::autograd_hook(VariableIndex index) {
     return;
   }
 
+  // Rebuilt bucket only if 1) it is first time rebuilt 2) unused_parameters_
```
Rebuilt bucket -> Rebuild bucket

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -38,6 +38,8 @@ namespace {
 constexpr char* GLOO_SOCKET_IFNAME_ENV = "GLOO_SOCKET_IFNAME";
 #endif
 
+constexpr int DEFAULT_BUCKET_BYTES_CAP = int(25 * 1024 * 1024);
```
The default value now lives in both Python and C++, do we need to consolidate them?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -8,10 +8,15 @@
 namespace c10d {
+
+constexpr int DEFAULT_FIRST_BUCKET_BYTES = int(1024 * 1024);
+constexpr int DEFAULT_BROADCAST_BUCKET_BYTES = int(250 * 1024 * 1024);
```
what's the reason for using 10X broadcast buckets byte sizes compared to default allreduce buckets?

zhaojuanmao replied:
distributed.py uses 10x broadcast buckets byte sizes in default, my understanding is because broadcast is cheaper than all reduce, so it should be fast enough to broadcast much larger buckets

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -8,10 +8,15 @@
 
 namespace c10d {
+
+constexpr int DEFAULT_FIRST_BUCKET_BYTES = int(1024 * 1024);
```
Share this with Python DDP?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +753,59 @@ void Reducer::finalize_backward() {
+  bucket_size_limits.push_back(DEFAULT_FIRST_BUCKET_BYTES);
+  bucket_size_limits.push_back(bucket_bytes_cap_);
+  rebuilt_bucket_indices = compute_bucket_assignment_by_size(
+      rebuilt_tensors_,
```
nit discussion: is it intentional for using `rebuilt_tensors_` instead of `rebuild_tensors_`?

zhaojuanmao replied:
feel rebuild_tensors_ means an action, but rebuilt_tensors_ means a noun, so using rebuilt_... for those variables, how do you think?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +753,59 @@ void Reducer::finalize_backward() {
+  int64_t broadcast_bucket_size = DEFAULT_BROADCAST_BUCKET_BYTES;
+  auto num_buckets = rebuilt_bucket_indices.size();
+  auto num_elem_per_bucket = rebuilt_bucket_indices[0].size();
+  int64_t total_size = num_buckets * num_elem_per_bucket;
```
is it guaranteed that all buckets have the same number of elements?

zhaojuanmao replied:
Good catch, let me double check

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +753,59 @@ void Reducer::finalize_backward() {
+  was_rebuilt_bucket_ = true;
+  rebuilt_tensors_.clear();
+  rebuilt_param_indices_.clear();
+  mutex_.unlock();
```
This is locked in the final callback installed by `mark_variable_ready`, right? Scattering the lock and unlock in two different functions might be error prone. Is there a way that we can apply the lock in a single place?

zhaojuanmao replied:
right, it is locked in final callback, that is why I unblock it here, otherwise initialize_buckets will hang, as initialize_buckets is trying to acquire lock to run. Let me think about how to consolidate them

zhaojuanmao replied:
had second thought, it is hard to consolidate locks here. initialize_buckets() is a public method that needs lock. it could be called in reducer constructor where no lock before calling, or could be called in rebuild_buckets where we have to unlock it before calling.

mrshenli replied:
Unresolve this for now, will come back to this later to discuss if there is a better to way to manage locks.

mrshenli replied:
Does the following make sense.

Since `rebuildBuckets` is only called in `finalize_backward`, and `finalize_backward` `mark_variable_ready`, can we move `rebuildBuckets` out of `finalize_backward` and directly call it in the lambda in `mark_variable_ready`? There we have directly control on the lock. Then we can also move `initialize_buckets` out from `rebuildBuckets`, sth like:

```c++
    // in mark_variable_ready
    torch::autograd::Engine::get_default_engine().queue_callback([=] {
      std::lock_guard<std::mutex> lock(this->mutex_);
      this->finalize_backward();
      auto rebuilt_bucket_indices = rebuildBuckets(...);
      lock.unlock();
      initialize_buckets(std::move(rebuilt_bucket_indices));
    });
```

zhaojuanmao replied:
sounds good!

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -186,11 +189,21 @@ class Reducer {
+  bool was_rebuilt_bucket_;
+  std::vector<at::Tensor> rebuilt_tensors_;
+  std::vector<int64_t> rebuilt_param_indices_;
+  int64_t bucket_bytes_cap_;
```
this can be const?

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -457,7 +477,6 @@ void Reducer::initialize_buckets(
             "that expect a sparse gradient.");
       }
     }
-
```
unintentional line delete?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -520,6 +539,7 @@ void Reducer::initialize_buckets(
           .intra_bucket_index = intra_bucket_index++,
       };
     }
+
```
unintentional new line?

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -186,11 +205,21 @@ class Reducer {
   std::vector<std::vector<int64_t>> backward_stats_;
+
+  // Following variables are to help build dynamic bucket order
+  bool was_rebuilt_bucket_;
```
should this be `has_rebuilt_buckets_`?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +750,75 @@ void Reducer::finalize_backward() {
   }
   local_used_maps_reduced_ = false;
+
+  // Rebuilt bucket if this is the first time to rebuilt
```
Rebuilt -> Rebuild
to rebuilt ->to rebuild?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -282,6 +287,21 @@ void Reducer::autograd_hook(VariableIndex index) {
+  // Also we only need to dump tensors and parameter indcies of one replica.
+  if (!was_rebuilt_bucket_ && unused_parameters_.empty() &&
+      index.replica_index == 0) {
+    rebuilt_tensors_.push_back(
```
call this `rebuilt_params_` to match the name of `rebuilt_param_indices_` below?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +750,75 @@ void Reducer::finalize_backward() {
+      process_group_, broadcast_tensor_device, broadcast_bucket_size);
+  broadcast_tensor.copy_(broadcast_tensor_device);
+  accessorIndex = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
```
Do we also need to broadcast number of buckets and size of each buckets? What if the mapping on each rank is different that results in different `num_buckets` and `bucket_size`?

zhaojuanmao replied:
you are right!

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -730,6 +750,75 @@ void Reducer::finalize_backward() {
+      rebuilt_param_indices_);
+
+  // For rebuilt bucket indices, it needs to be synced across all ranks.
+  // broadcast the newly rebuilt bucket indices from rank 0 in default.
```
broadcast -> Broadcast

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -282,6 +287,21 @@ void Reducer::autograd_hook(VariableIndex index) {
+  // Here, we just dump tensors and their parameter indices into
+  // rebuilt_tensors_ and rebuilt_param_indices_, and then at the end of
+  // finalize_backward(), buckets will be rebuilt based on rebuilt_tensors_
+  // and rebuilt_param_indices_, and then will be broadcasted and intialized.
```
nit: spelling

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -282,6 +287,21 @@ void Reducer::autograd_hook(VariableIndex index) {
+  // rebuilt_tensors_ and rebuilt_param_indices_, and then at the end of
+  // finalize_backward(), buckets will be rebuilt based on rebuilt_tensors_
+  // and rebuilt_param_indices_, and then will be broadcasted and intialized.
+  // Also we only need to dump tensors and parameter indcies of one replica.
```
nit: spelling

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

mrshenli commented:
Test failures are real. E.g.,

```
May 12 01:40:47 ======================================================================
May 12 01:40:47 ERROR [0.109s]: test_ignored_output_with_unused_parameters (__main__.DistributedDataParallelTest)
May 12 01:40:47 ----------------------------------------------------------------------
May 12 01:40:47 Traceback (most recent call last):
May 12 01:40:47   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 214, in wrapper
May 12 01:40:47     self._join_processes(fn)
May 12 01:40:47   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 316, in _join_processes
May 12 01:40:47     self._check_return_codes(elapsed_time)
May 12 01:40:47   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 349, in _check_return_codes
May 12 01:40:47     raise RuntimeError(error)
May 12 01:40:47 RuntimeError: Processes 0 1 exited with error code 10
May 12 01:40:47 
May 12 01:40:47 ======================================================================
May 12 01:40:47 ERROR [0.108s]: test_sparse_gradients (__main__.DistributedDataParallelTest)
May 12 01:40:47 ----------------------------------------------------------------------
May 12 01:40:47 Traceback (most recent call last):
May 12 01:40:47   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 214, in wrapper
May 12 01:40:47     self._join_processes(fn)
May 12 01:40:47   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 316, in _join_processes
May 12 01:40:47     self._check_return_codes(elapsed_time)
May 12 01:40:47   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 349, in _check_return_codes
May 12 01:40:47     raise RuntimeError(error)
May 12 01:40:47 RuntimeError: Processes 0 1 exited with error code 10
```

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -210,7 +210,8 @@ class DistributedDataParallel(Module):
     """
     def __init__(self, module, device_ids=None,
                  output_device=None, dim=0, broadcast_buffers=True,
-                 process_group=None, bucket_cap_mb=25,
+                 process_group=None,
+                 bucket_cap_mb=25,
```
I tried to use exported constant from C++, but it failed on windows build. windows "import torch" will import torch.nn.parallel modules that include importing DistributedDataParallel. but windows do not have distributed package, and will complain exported constant is not available. 

I also tried to import DistributedDataParallel only if distributed.is_available() in torch.nn._init_, but that break a lot of other projects. 

I think it should be fine to let this constant in both python and C++? as they are defaults for different constructors. one is for DistributedDataParallel, one is for pybinded reducer. DistributedDataParallel and reducer are usually used together, we always pass bucket_cap_mb to reducer from DistributedDataParallel. 

How do you think?

mrshenli replied:
I see, yes, in this case, let's keep it this way. This might be another reason that we should move DDP to `torch.distributed`

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -743,6 +746,14 @@ They are used in specifying strategies for reduction collectives, e.g.,
       },
       py::call_guard<py::gil_scoped_release>());
 
+  module.def("_default_first_bucket_bytes", []() {
+    return ::c10d::DEFAULT_FIRST_BUCKET_BYTES;
+  });
+
+  module.def("_default_broadcast_bucket_bytes", []() {
+    return ::c10d::DEFAULT_BROADCAST_BUCKET_BYTES;
+  });
```
Does these two have to be methods? Can we use `module.attr` instead? E.g.

```c++
  module.attr("_DEFAULT_FIRST_BUCKET_BYTES") =
      py::cast(kDefaultBroadcastBucketBytes);
```

For consistency, we use capital letters in constants, and use `kCamelName` in C++ to name it.  One example is `_DEFAULT_NUM_SEND_RECV_THREADS` in `rpc/init.cpp`

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -759,7 +883,8 @@ inline bool operator==(const BucketKey& lhs, const BucketKey& rhs) {
 std::vector<std::vector<size_t>> compute_bucket_assignment_by_size(
```
nit: This function is called in two places: 1) DDP ctor and 2) when rebuilding mapping. Its behaviors in these two situations are different and are governed by the value in `tensor_indices` arg. I wonder if we can split this into two methods to make it clear? E.g, we can do the following:

1. Add a new `default_bucket_assignment_by_size` which does not take a `tensor_indices` vector
2. Modify `compute_bucket_assignment_by_size` to let it always expect `tensor_indices` to contain the real indices
3. `default_bucket_assignment_by_size` calls into `compute_bucket_assignment_by_size` by providing a default vector, where `tensor_indices[i] = i`. As this is only called in ctor, the overhead of creating a new vector should be OK.
4. Expose `default_bucket_assignment_by_size` to Python.

zhaojuanmao replied:
this sounds a good idea, but I checked default_bucket_assignment_by_size is also called by broadcast_coalesced(), the broadcast_coalesced() is called in forward() per iteration if broadcast_buffer is enabled. that means every iteration needs to create a tensor_indices vector, it may be costly. 

On the other hand, I feel like keeping tensor_indices to be empty in default in compute_bucket_assignment_by_size() is probably OK, it implicitly means tensor_indices[i] = i, which is default bucket assignment. let me make comments more clear?

zhaojuanmao replied:
default_bucket_assignment_by_size is called in broadcast_coalesced as well, broadcast_coalesced is called per iteration, adding tensor_indices every iteration may be costly.

on the other hand, I feel like it is not too bad to make tensor_indices to be empty in default, representing  tensor_indices[i] = i. I'm making comments more clear, please let me know whether it makes sense to you?

mrshenli replied:
I see, you mean the buffer sync in Python `_sync_params` is calling into this? In this case, let's leave it as is.

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -271,6 +274,21 @@ void Reducer::autograd_hook(VariableIndex index) {
 
+  // Rebuild bucket only if 1) it is the first time to rebuild bucket 2)
+  // unused_parameters_ is empty, currently it does not support when there are
+  // unused parameters 3) this backward pass needs to run all reduce. Here, we
```
all reduce -> allreduce

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -375,6 +393,14 @@ void Reducer::mark_variable_ready(VariableIndex index) {
     torch::autograd::Engine::get_default_engine().queue_callback([=] {
       std::lock_guard<std::mutex> lock(this->mutex_);
```
Since we need to unlock below, can we use a `unique_lock` instead, make both unlock explicit?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -375,6 +393,14 @@ void Reducer::mark_variable_ready(VariableIndex index) {
+        auto rebuilt_bucket_indices = rebuildBuckets();
+        // Unlock before initialize_buckets() as initialize_buckets() requires a
+        // lock, it could result in self deadlock without unlocking here.
+        mutex_.unlock();
```
if using `unique_lock`, here can be changed to `lock.unlock()`

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -375,6 +393,14 @@ void Reducer::mark_variable_ready(VariableIndex index) {
+        // lock, it could result in self deadlock without unlocking here.
+        mutex_.unlock();
+        initialize_buckets(std::move(rebuilt_bucket_indices));
+      }
```
if using `unique_lock`, need an `else` clause here that does the unlock

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -13,6 +13,10 @@
 
 namespace c10d {
 
+constexpr int DEFAULT_FIRST_BUCKET_BYTES = int(1024 * 1024);
+constexpr int DEFAULT_BUCKET_BYTES_CAP = int(25 * 1024 * 1024);
+constexpr int DEFAULT_BROADCAST_BUCKET_BYTES = int(250 * 1024 * 1024);
```
Let's stay consistent with other parts of our code base and use `kDefaultBroadcastBucketBytes` instead.

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -108,6 +113,20 @@ class Reducer {
+  // the buckets
+  void sync_bucket_indices(std::vector<std::vector<size_t>>& bucket_indices);
+  // Rebuild buckets based on rebuilt_params_ and rebuilt_param_indices_
+  // TODO this function makes broadcast communication call and
```
Let's add an issue to track this.

zhaojuanmao replied:
created issue #38543

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -729,6 +755,104 @@ void Reducer::finalize_backward() {
+  int64_t broadcast_bucket_size = DEFAULT_BROADCAST_BUCKET_BYTES;
+
+  auto num_buckets = bucket_indices.size();
+  std::vector<size_t> bucket_sizes;
```
add a `bucket_sizes.reserve(num_buckets)`?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -729,6 +755,104 @@ void Reducer::finalize_backward() {
+  // Group indices and num_bucket together into indices_tensor
+  // Broadcast this tensor first, as its size is equal among all processes
+  auto indices_tensor = at::empty({total_size + 1}, at::kInt);
+  auto indices_accessor = indices_tensor.accessor<int, 1>();
```
Question, do we have to use the accessor here? I recall hearing some update from C++ team that C++ tensor indexing support is ready?

zhaojuanmao replied:
I was recommended to use accessor for efficiency last year when I worked on data loading API. I did not find C++ tensor indexing documentation, is it a new feature?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -729,6 +755,104 @@ void Reducer::finalize_backward() {
+  // it can only broadcast device tensors.
+  auto indices_tensor_device = at::empty({total_size + 1}, options);
+  indices_tensor_device.copy_(indices_tensor, true);
+  broadcast_coalesced(
```
Since `indices_tensor_device` is a single tensor, we don't need "coalesced" here? Looking at its implementation in `comm.cpp`, it is calling `compute_bucket_assignment_by_size` again and have additional unnecessary for/while loops. Can we directly call broadcast and wait on the result here?

zhaojuanmao replied:
sounds good! let me try that

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -729,6 +755,104 @@ void Reducer::finalize_backward() {
+  }
+  auto bucket_sizes_tensor_device = at::empty({(int64_t)num_buckets}, options);
+  bucket_sizes_tensor_device.copy_(bucket_sizes_tensor, true);
+  broadcast_coalesced(
```
ditto

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -729,6 +755,104 @@ void Reducer::finalize_backward() {
+    for (size_t j = 0; j < bucket_size; j++) {
+      bucket.push_back(indices_accessor[indices_accessor_Index++]);
+    }
+    bucket_indices.push_back(bucket);
```
emplace_back + std::move?

mrshenli reviewed (commented):
Thanks for addressing comments. This is getting better and better! I left some additional minor comments inline.

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -9,6 +9,7 @@
 #include <torch/csrc/autograd/functions/accumulate_grad.h>
 #include <torch/csrc/autograd/profiler.h>
 #include <torch/csrc/autograd/utils/lambda_post_hook.h>
+#include <torch/csrc/distributed/c10d/comm.h>
```
Do we still need this header after removing `broadcast_coalesced` in this file?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -729,6 +758,107 @@ void Reducer::finalize_backward() {
+  // After syncing up rebuilt bucket indices, initialize buckets for reducer.
+  sync_bucket_indices(rebuilt_bucket_indices);
+
+  has_rebuilt_bucket_ = true;
```
nit: let's add an assert in the beginning of this function to make sure that `has_rebuilt_bucket_ ` is indeed false.

mrshenli reviewed (approved):
LGTM! Thanks for adding this!

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "implement dynamic bucket order in DDP"


bucket order is rebuilt dynamically in the first reduction backward pass when find_unused_parameters = false

Differential Revision: [D20128537](https://our.internmc.facebook.com/intern/diff/D20128537/)

[ghstack-poisoned]
